geetools/ee_profiler.py

```diff
@@ -0,0 +1,94 @@
+"""A profiler context manager for Earth Engine Python API."""
+from __future__ import annotations
+
+import io
+import re
+
+import ee
+
+from .accessors import _register_extention
+
+
+@_register_extention(ee)
+class Profiler:
+    """A profiler context manager for Earth Engine Python API.
+
+    Examples:
+        .. code-block:: python
+
+            import ee, geetools
+
+            ee.Initialize()
+
+            with ee.Profiler() as p:
+                ee.Number(3.14).add(0.00159).getInfo()
+                res = p.profile
+    """
+
+    _output_capture: io.StringIO | None = None
+    "The output of the profiler."
+
+    _profile_context: ee.profilePrinting | None = None
+    "The raw profile context."
+
+    profile: dict | None = None
+    "The profile data as a dictionary."
+
+    def __enter__(self):
+        """Enter the context manager."""
+        self._output_capture = io.StringIO()
+        self._profile_context = ee.profilePrinting(destination=self._output_capture)
+        self._profile_context.__enter__()
+        return self
+
+    def __exit__(self, *args):
+        """Exit the context manager."""
+        self._profile_context.__exit__(*args)
+
+        # Check if there's anything captured
+        profile_output = self._output_capture.getvalue()
+        if profile_output:
+            self.profile = self.to_dict(profile_output)
+        else:
+            self.profile = None  # Handle the case where no output is captured
+            print("Warning: No profile output was captured.")
+
+        self._output_capture.close()
+
+    def memory(self, mem_str: str) -> int:
+        """Transform a memory string to an integer."""
+        mapping = {"": 1, "k": 3, "M": 6, "G": 9, "T": 12}
+
+        # Match numbers with optional multipliers (k, M, etc.)
+        # and apply the multiplier to the number
+        match = re.match(r"([\d.]+)([kMGT]?)", mem_str)
+        if match is None:
+            raise ValueError(f"Invalid memory string: {mem_str}")
+
+        number, multiplier = float(match.group(1)), match.group(2)
+
+        return int(number * 10 ** mapping[multiplier])
+
+    def to_dict(self, input: str) -> dict:
```

Comment:
maybe this method should not be exposed directly to user, since it's used internally, what about `._to_dict`?

Reply:
solved in https://github.com/gee-community/geetools/pull/358/commits/5fe2eef2f0de020a78c5a36b1f3d8d23645a501f